src/librustc_codegen_ssa/mir/place.rs

```diff
@@ -240,38 +243,49 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     layout::Int(_, signed) => !discr_scalar.is_bool() && signed,
                     _ => false
                 };
-                bx.intcast(lldiscr, cast_to, signed)
+                bx.intcast(encoded_discr.immediate(), cast_to, signed)
             }
             layout::DiscriminantKind::Niche {
                 dataful_variant,
                 ref niche_variants,
                 niche_start,
             } => {
-                let niche_llty = bx.cx().immediate_backend_type(discr.layout);
-                if niche_variants.start() == niche_variants.end() {
+                let niche_llty = bx.cx().immediate_backend_type(encoded_discr.layout);
+                let encoded_discr = encoded_discr.immediate();
+                let (is_niche, niche_discr) = if niche_variants.start() == niche_variants.end() {
```

Comment:
Personally, I find the use of a tuple pretty hard to read in code like this, because of how far away the tuple-constructing expressions are.

Did you consider doing:

```rust
let is_niche;
let niche_discr;
if  ... { ... is_niche = A; niche_discr = B; } else { ... is_niche = C; niche_discr = D; }
```
?

Reply:
That's fair, but I don't think the second commit keeps a tuple anyway.

Reply:
Ah you are right